Run test_torchinductor*.py with implicit_fallbacks=False

fdrocha opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94206
* __->__ #94039

This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Commit:
Run test_torchinductor_opinfo.py with implicit_fallbacks=False

This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

[ghstack-poisoned]

lezcano reviewed (approved):
Fair enough.

Note that many of these may be removed by registering the corresponding decomposition with the `Autograd` key. It'd be good to use this list to start adding the relevant decomps.

fdrocha commented:
> Fair enough.
> 
> Note that many of these may be removed by registering the corresponding decomposition with the `Autograd` key. It'd be good to use this list to start adding the relevant decomps.

Yes, that's actually how I ended up doing this, as the precursor to doing that.

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -87,7 +87,8 @@
 config.triton.autotune_pointwise = False  # too slow
-
+# make sure we get errors if we don't have explicit fallback
+config.implicit_fallbacks = False
```
Let's move this (and the one above) to a decorator:
```
@config.patch(implicit_fallbacks=False)
```

You could do something like:
https://github.com/pytorch/pytorch/blob/6c555b29a8cf23e99c03887b75af22bf43f52d41/torch/_dynamo/test_case.py#L44-L55

fdrocha replied:
Thanks, it's done

ngimel commented on torch/_inductor/lowering.py:
```diff
@@ -1225,6 +1225,211 @@ def apply_constraint(arg, fx_arg):
+
+make_fallback(torch.ops.prims.amin)
+make_fallback(torch.ops.prims.gcd.default)
+make_fallback(torch.ops.prims.shift_left.default)
```
hm, I wonder how did these appear, we must be using some questionable refs

ngimel replied:
I guess it's comming from unwanted `register_decomposition` for an inplace version

jansel replied:
We have lowerings for the aten versions of some of those.  Perhaps we shouldn't actually be falling back.

ngimel replied:
Yeah we have or could create trivial pointwise lowerings for a lot of those, but we shouldn't be seeing prims in the graphs we are getting.

fdrocha replied:
Yes, meant to ask you guys about this. I've been looking into it and it's a couple of different things:

- For `shift_left`, I managed to get rid of the prim fallback, just had to add a missing implementation to `codegen/common.py`
- For `shift_right`, same story, but the naive implementation gives the wrong results on cuda somehow. Need to look into this some more.
- For `gcd` I think we just need to actually implement the prim in inductor, it's not there. Alternatively, we can drop the gcd decomposition from the inductor list and register a fallback for `aten.gcd` instead, probably better.
- For `amin`, the issue is that the reference for `linalg.norm` is calling `refs.amin` instead of `torch.amin`. If you fix that this issue no longer comes up, but then some mypy checks fail - need to sort that out.

ngimel replied:
Ok, let's fix linalg.norm decomp to call torch.amin, mypy shouldn't be too hard, and even if it is, leave a comment to ignore it?
As for shift_right/left, yeah we can have lowering, I still wonder how did prim get into the graph.

fdrocha replied:
Should be all set.

ngimel commented on torch/_inductor/lowering.py:
```diff
@@ -1225,6 +1225,211 @@ def apply_constraint(arg, fx_arg):
+make_fallback(aten.expand_copy)
+make_fallback(aten.zeros)
+
+make_fallback(torch.ops.prims.amin)
```
same here

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/common.py:
```diff
@@ -123,6 +123,14 @@ def bitwise_or(x, y):
+    def bitwise_left_shift(x, y):
+        return f"{ExprPrinter.paren(x)} << {ExprPrinter.paren(y)}"
+
+    #    @staticmethod
```
Why is the commented out?

fdrocha replied:
There is a bug with right shifts in Triton that leads to a test failure.
I just submitted a PR fixing it (https://github.com/openai/triton/pull/1153), which will probably not take long to land, but since it is non-trivial to advance our triton pin, I put some comments in the code and registered a fallback to the aten right shift for now.

jansel replied:
Mention this in a comment

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1424,9 +1424,8 @@ def apply_constraint(arg, fx_arg):
-make_fallback(torch.ops.prims.amin)
 make_fallback(torch.ops.prims.gcd.default)
-make_fallback(torch.ops.prims.shift_left.default)
+make_fallback(torch.ops.prims.amin)
```
map this to aten.amin (which is supported)

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1222,6 +1222,210 @@ def apply_constraint(arg, fx_arg):
+make_fallback(aten.upsample_trilinear3d)
+make_fallback(aten.upsample_linear1d_backward)
+make_fallback(aten.upsample_trilinear3d_backward)
+make_fallback(aten.acos)
```
should be easy to support this one

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1222,6 +1222,210 @@ def apply_constraint(arg, fx_arg):
+make_fallback(aten.upsample_linear1d_backward)
+make_fallback(aten.upsample_trilinear3d_backward)
+make_fallback(aten.acos)
+make_fallback(aten.acosh)
```
support this

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1222,6 +1222,210 @@ def apply_constraint(arg, fx_arg):
+make_fallback(aten.avg_pool3d_backward)
+make_fallback(aten.binary_cross_entropy_backward)
+make_fallback(aten.bitwise_or_)
+make_fallback(aten._cdist_backward)
+make_fallback(aten.diagonal_backward)
+make_fallback(aten._embedding_bag_dense_backward)
+make_fallback(aten.fractional_max_pool2d_backward)
+make_fallback(aten.fractional_max_pool3d_backward)
+make_fallback(aten.hardshrink_backward)
+make_fallback(aten.huber_loss_backward)
+make_fallback(aten._linalg_check_errors)
+make_fallback(aten.log_sigmoid_backward)
+make_fallback(aten.max_pool3d_with_indices_backward)
+make_fallback(aten.multilabel_margin_loss_backward)
+make_fallback(aten.multi_margin_loss_backward)
+make_fallback(aten._pdist_backward)
+make_fallback(aten._prelu_kernel_backward)
+make_fallback(aten.reflection_pad1d_backward)
+make_fallback(aten.replication_pad1d_backward)
+make_fallback(aten.smooth_l1_loss_backward)
+make_fallback(aten.soft_margin_loss_backward)
+make_fallback(aten.softshrink_backward)
+make_fallback(aten.squeeze_copy)
+make_fallback(aten.linalg_pinv.atol_rtol_tensor)
+make_fallback(aten.segment_reduce.default)
+make_fallback(aten._segment_reduce_backward.default)
+make_fallback(aten.angle)
+make_fallback(aten.cholesky_inverse)
+make_fallback(aten.cholesky_solve)
+make_fallback(aten._fft_r2c)
+make_fallback(aten.histogram.bin_ct)
+make_fallback(aten._histogramdd_bin_edges.default)
+make_fallback(aten._histogramdd_from_bin_cts.default)
+make_fallback(aten.index_reduce)
+make_fallback(aten.masked_scatter)
+make_fallback(aten.to_sparse)
+make_fallback(aten.triangular_solve)
+make_fallback(aten.expand_copy)
+make_fallback(aten.zeros)
+make_fallback(aten.gcd.default)
```
I see lots of simple pointwise ops that could be supported with a few lines of code and also variants of things already supported here. We should just add support for the easy ones.

lezcano replied:
may be better to do that in a follow-up PR in this stack not to clutter this one too much.

fdrocha replied:
Yes, I'm working on this. I was going to do it as a second PR in same `ghstack` stack but LMK if you think it's better to do all in same PR.

jansel reviewed (changes requested):
Looks like some changes to fbgemm snuck in by mistake?

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

fdrocha commented:
> Looks like some changes to fbgemm snuck in by mistake?

doh, fixed

ngimel commented on torch/_refs/linalg/__init__.py:
```diff
@@ -70,11 +70,13 @@ def vector_norm(
     # Checks
     check_fp_or_complex(x.dtype, "linalg.vector_norm")
 
+    if dim is None:
```
would it be better to avoid `flatten` and set `dim=range(ndim)`? Flatten complicates the indexing if input was permuted.

fdrocha replied:
Yes, flatten was a bad idea here. I got rid of this in a subsequent commit already.

fdrocha replied:
Sorry, previous comment wasn't true when I said it, but it is now. (Very strange, I was positive I committed that change already)

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "Run test_torchinductor*.py with implicit_fallbacks=False"


This way it errors out for ops that don't have decomps and
requires you to add explicit fallbacks to lowering.py

Turns out there are a lot, and this commit adds them as well.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]
